File: src/main/java/at/hannibal2/skyhanni/config/features/About.java
package at.hannibal2.skyhanni.config.features;

import at.hannibal2.skyhanni.config.FeatureToggle;
import at.hannibal2.skyhanni.features.misc.update.ConfigVersionDisplay;
import at.hannibal2.skyhanni.utils.OSUtils;
import com.google.gson.annotations.Expose;
import io.github.notenoughupdates.moulconfig.annotations.Accordion;
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean;
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorButton;
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorDropdown;
import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
import io.github.notenoughupdates.moulconfig.observer.Property;

public class About {

    @ConfigOption(name = "Current Version", desc = "This is the SkyHanni version you are currently running")
    @ConfigVersionDisplay
    public transient Void currentVersion = null;

    @ConfigOption(name = "Check for Updates", desc = "Automatically check for updates on each startup")
    @Expose
    @ConfigEditorBoolean
    public boolean autoUpdates = true;

    @ConfigOption(name = "Auto Updates", desc = "Automatically download new version on each startup")
    @Expose
    @FeatureToggle
    @ConfigEditorBoolean
    public boolean fullAutoUpdates = false;

    @ConfigOption(name = "Update Stream", desc = "How frequently you want updates for SkyHanni")
    @Expose
    @ConfigEditorDropdown
    public Property<UpdateStream> updateStream = Property.of(UpdateStream.RELEASES);


    @ConfigOption(name = "Used Software", desc = "Information about used software and licenses")
    @Accordion
    @Expose
    public Licenses licenses = new Licenses();

    public enum UpdateStream {
        NONE("None", "none"),
        BETA("Beta", "pre"),
        RELEASES("Full", "full");

        private final String label;
        private final String stream;

        UpdateStream(String label, String stream) {
            this.label = label;
            this.stream = stream;
        }

        public String getStream() {
            return stream;
        }

        @Override
        public String toString() {
            return label;
        }
    }

    public static class Licenses {

        @ConfigOption(name = "MoulConfig", desc = "MoulConfig is available under the LGPL 3.0 License or later version")
        @ConfigEditorButton(buttonText = "Source")
        public Runnable moulConfig = () -> OSUtils.openBrowser("https://github.com/NotEnoughUpdates/MoulConfig");

        @ConfigOption(name = "NotEnoughUpdates", desc = "NotEnoughUpdates is available under the LGPL 3.0 License or later version")
        @ConfigEditorButton(buttonText = "Source")
        public Runnable notEnoughUpdates = () -> OSUtils.openBrowser("https://github.com/NotEnoughUpdates/NotEnoughUpdates");

        @ConfigOption(name = "Forge", desc = "Forge is available under the LGPL 3.0 license")
        @ConfigEditorButton(buttonText = "Source")
        public Runnable forge = () -> OSUtils.openBrowser("https://github.com/MinecraftForge/MinecraftForge");

        @ConfigOption(name = "LibAutoUpdate", desc = "LibAutoUpdate is available under the BSD 2 Clause License")
        @ConfigEditorButton(buttonText = "Source")
        public Runnable libAutoUpdate = () -> OSUtils.openBrowser("https://github.com/nea89o/libautoupdate");

        @ConfigOption(name = "Mixin", desc = "Mixin is available under the MIT License")
        @ConfigEditorButton(buttonText = "Source")
        public Runnable mixin = () -> OSUtils.openBrowser("https://github.com/SpongePowered/Mixin/");

        @ConfigOption(name = "DiscordIPC", desc = "DiscordIPC is available under the Apache License 2.0")
        @ConfigEditorButton(buttonText = "GitHub")
        public Runnable discordRPC = () -> OSUtils.openBrowser("https://github.com/jagrosh/DiscordIPC");
    }
}

File: src/main/java/at/hannibal2/skyhanni/config/features/About.kt
package at.hannibal2.skyhanni.config.features

import at.hannibal2.skyhanni.features.misc.update.ConfigVersionDeprecatedDisplay
import at.hannibal2.skyhanni.features.misc.update.ConfigVersionDisplay
import at.hannibal2.skyhanni.utils.OSUtils.openBrowser
import at.hannibal2.skyhanni.utils.system.ModVersion
import com.google.gson.annotations.Expose
import io.github.notenoughupdates.moulconfig.annotations.Accordion
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorButton
import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorDropdown
import io.github.notenoughupdates.moulconfig.annotations.ConfigOption
import io.github.notenoughupdates.moulconfig.annotations.SearchTag
import io.github.notenoughupdates.moulconfig.observer.Property

class About {
    @ConfigOption(name = "", desc = "")
    @ConfigVersionDeprecatedDisplay
    @Transient
    var deprecatedVersionWarning: Unit? = null

    @Suppress("unused")
    @ConfigOption(name = "Current Version", desc = "This is the SkyHanni version you are currently running")
    @ConfigVersionDisplay
    @SearchTag("check download update")
    @Transient
    var currentVersion: Unit? = null

    @ConfigOption(name = "Update Stream", desc = "How frequently you want updates for SkyHanni")
    @Expose
    @ConfigEditorDropdown
    val updateStream: Property<UpdateStream> = Property.of(UpdateStream.forInstalledVersion())

    @Suppress("unused")
    @ConfigOption(name = "Used Software", desc = "Information about used software and licenses")
    @Accordion
    @Expose
    val licenses: Licenses = Licenses()

    enum class UpdateStream(private val label: String) {
        BETA("Beta"),
        RELEASES("Full");

        override fun toString() = label

        companion object {
            /**
             * The stream matching the installed version, used as the default on a clean install.
             *
             * Reads [ModVersion.installed] and not `SkyHanniMod.isBetaVersion`, because the config is also
             * constructed while `SkyHanniMod` is still initialising, where that property may be unset.
             */
            fun forInstalledVersion(): UpdateStream = if (ModVersion.installed.isBeta) BETA else RELEASES
        }
    }

    class Licenses {
        @ConfigOption(name = "MoulConfig", desc = "MoulConfig is available under the LGPL 3.0 License or later version")
        @ConfigEditorButton(buttonText = "Source")
        val moulConfig: Runnable = Runnable { openBrowser("https://github.com/NotEnoughUpdates/MoulConfig") }

        @ConfigOption(name = "NotEnoughUpdates-REPO", desc = "NotEnoughUpdates-REPO is available under the MIT License")
        @ConfigEditorButton(buttonText = "Source")
        val notEnoughUpdatesRepo: Runnable = Runnable { openBrowser("https://github.com/NotEnoughUpdates/NotEnoughUpdates-REPO") }

        @ConfigOption(name = "Fabric Loader", desc = "Fabric Loader is available under the Apache-2.0 license")
        @ConfigEditorButton(buttonText = "Source")
        val fabricLoader: Runnable = Runnable { openBrowser("https://github.com/FabricMC/fabric-loader") }

        @ConfigOption(name = "Fabric API", desc = "Fabric API is available under the Apache-2.0 license")
        @ConfigEditorButton(buttonText = "Source")
        val fabricApi: Runnable = Runnable { openBrowser("https://github.com/FabricMC/fabric-api") }

        @ConfigOption(name = "Mixin", desc = "Mixin is available under the MIT License")
        @ConfigEditorButton(buttonText = "Source")
        val mixin: Runnable = Runnable { openBrowser("https://github.com/FabricMC/Mixin") }

        @ConfigOption(name = "MixinExtras", desc = "MixinExtras is available under the MIT License")
        @ConfigEditorButton(buttonText = "Source")
        val mixinExtras: Runnable = Runnable { openBrowser("https://github.com/LlamaLad7/MixinExtras") }
    }
}
